mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala

```diff
@@ -114,10 +114,10 @@ final class QuantileDiscretizer @Since("1.6.0") (@Since("1.6.0") override val ui
     splits(0) = Double.NegativeInfinity
     splits(splits.length - 1) = Double.PositiveInfinity
 
-    val distinctSplits = splits.distinct
+    val distinctSplits = splits.filter(!_.isNaN).distinct
     if (splits.length != distinctSplits.length) {
-      log.warn(s"Some quantiles were identical. Bucketing to ${distinctSplits.length - 1}" +
-        s" buckets as a result.")
+      log.warn(s"Bucketizing to ${distinctSplits.length - 1} buckets if some quantiles" +
+        s" were identical; or ${distinctSplits.length} buckets if NaN value existed")
```

Comment:
I think this is now a bit confusing, since it's reporting different things based on state that isn't logged for the user. If it's hard just say "bucketing to fewer buckets" as before at this stage.